internal/command/jsonformat/plan_test.go

```diff
@@ -77,6 +77,299 @@ and found no differences, so no changes are needed.
+								"value": "Hello, World!",
+							}),
+							Importing: true,
+						},
+					},
+				},
+			},
+			output: `
+Terraform used the selected providers to generate the following execution
+plan. Resource actions are indicated with the following symbols:
+  & import
+
+Terraform will perform the following actions:
+
+  # test_resource.resource will be imported and then destroyed
+  & resource "test_resource" "resource" {
+      - id    = "1D5F5E9E-F2E5-401B-9ED5-692A215AC67E" -> null
+      - value = "Hello, World!" -> null
+    }
+
+Plan: 0 to add, 1 to import, 0 to change, 0 to destroy.
+`,
+		},
+		"import_and_read": {
+			plan: Plan{
+				ResourceChanges: []jsonplan.ResourceChange{
+					{
+						Address:      "test_resource.resource",
+						Mode:         "managed",
+						Type:         "test_resource",
+						Name:         "resource",
+						ProviderName: "test",
+						Change: jsonplan.Change{
+							Actions: []string{"read"},
+							Before: marshalJson(t, map[string]interface{}{
+								"id":    "1D5F5E9E-F2E5-401B-9ED5-692A215AC67E",
+								"value": "Hello, World!",
+							}),
+							After: marshalJson(t, map[string]interface{}{
+								"id":    "1D5F5E9E-F2E5-401B-9ED5-692A215AC67E",
+								"value": "Hello, World!",
+							}),
+							Importing: true,
+						},
+					},
+				},
+			},
+			output: `
+Terraform used the selected providers to generate the following execution
+plan. Resource actions are indicated with the following symbols:
+  & import
+
+Terraform will perform the following actions:
+
+  # test_resource.resource will be imported and then read during apply
+  & resource "test_resource" "resource" {
+      & id    = "1D5F5E9E-F2E5-401B-9ED5-692A215AC67E"
+      & value = "Hello, World!"
+    }
+
+Plan: 0 to add, 1 to import, 0 to change, 0 to destroy.
+`,
+		},
+		"import_and_replace": {
+			plan: Plan{
+				ResourceChanges: []jsonplan.ResourceChange{
+					{
+						Address:      "test_resource.resource",
+						Mode:         "managed",
+						Type:         "test_resource",
+						Name:         "resource",
+						ProviderName: "test",
+						Change: jsonplan.Change{
+							Actions: []string{"create", "delete"},
+							Before: marshalJson(t, map[string]interface{}{
+								"id":    "1D5F5E9E-F2E5-401B-9ED5-692A215AC67E",
+								"value": "Hello, World!",
+							}),
+							After: marshalJson(t, map[string]interface{}{
+								"id":    "9794FB1F-7260-442F-830C-F2D450E90CE3",
+								"value": "Hello, World!",
+							}),
+							ReplacePaths: marshalJson(t, [][]string{{"id"}}),
+							Importing:    true,
+						},
+						ActionReason: "",
+					},
+				},
+			},
+			output: `
+Terraform used the selected providers to generate the following execution
+plan. Resource actions are indicated with the following symbols:
+  & import
+
+Terraform will perform the following actions:
+
+  # test_resource.resource must be imported and then replaced
+  & resource "test_resource" "resource" {
```

Comment:
This is a little trickier than the other cases because the replacement symbol carried extra information for the user. I'm not sure how to best format that data here, but do we want to try and convey the `-/+` vs `+/-` somehow? Maybe it's a small enough case that it doesn't really matter in practice.

Reply:
We are kinda running out of UI space here for all of these different ideas. Particularly with importing being an additional action alongside what we traditionally think of as the "action", it does seem a shame to effectively merge them back together again in the UI.

I wonder if perhaps we should not try to combine these quite so much and instead present the import as a separate action adjacent to the main action, if any...

```
  # test_resource.resource will first import an existing object
  & import {
      id = "whatever-id-value-in-config"
    }

  # test_resource.resource must be replaced after importing
  # (the usual explanation note)
  # Warning: this will destroy the imported object
-/+ resource "test_resource" "resource" {
      # (the usual stuff here)
    }
```

I imagine these two clauses always appearing adjacent to one another so that it's easier to see them as related. The message above the second clause is extended with "after importing" to try to reinforce the connection with and ordering relative to the import clause listed above.

This also makes the importing stand out more in the plan and makes the final resolved ID visible, which might help folks in cases where the ID is populated from a non-constant expression. Since importing is an occasional one-off operation rather than a routine operation, and that users are responsible for ensuring that they aren't violating Terraform's assumptions by importing the same object twice, I think some additional UI chatter is justified.

For the case where there aren't any additional changes required after importing (that is: when the main action is "no op") I might suggest a rendering similar to the above but without any action symbol at all next to the resource block, which is how we typically represent that something is happening in Terraform that won't directly affect the real infrastructure. In that case the comment above could be something like this:

```
  # test_resource.resource will first import an existing object
  & import {
      id = "whatever-id-value-in-config"
    }

  # test_resource.resource requires no changes after importing
    resource "test_resource" "resource" {
      # (the usual stuff here)
    }
```

Reply:
I do like this idea, but I wonder if maybe this shouldn't be the renderer's decision? Maybe Terraform itself could split these out as actual changes - that way they would appear in the json plan and the streamed logs as well. I'm not sure

We'll need to modify the renderers API to include accepting the config for this as well, but that's definitely possible.

I do think the main thing here is thinking about whether the two actions (import and then whatever) should be actually be part of the plan or whether this is just something the human readable plan renderer will do to make things nice.

Reply:
Importing itself isn't a "change" in the same way that the other actions are a change. Importing is in a sense just a modified version of the `ReadResource` call, so the fact that a resource was imported ends up being this meta-attribute about it in the planned change.

We could of course model this in other ways, but there are a number of things which already assume that there can only be a single change for each instance, so having separate import changes could end up being more complicated.

Reply:
For what its worth, I was imagining the two "blocks" I illustrated as being a single unit as far as the render is concerned... although we present it as two blocks in the UI -- because that echoes how it's written in the configuration -- it's still really just a single indivisible change. It's a UI decision how to present it, and our UI decisions typically echo how it looks in configuration just so it seems more familiar to a human operator.

However, I also don't have strong feelings about how it's represented internally. If representing importing as a separate action entirely ends up being easier then that does seem defensible, although the UI would then still need some special logic to notice that there's both an import and another action for the same resource instance and group them together for rendering as a single object, so I dunno if actually ends up being materially different in the UI layer anyway. :man_shrugging: